token.go

```diff
@@ -5,218 +5,262 @@
+func (p orderedPartitioner) Hash(partitionKey []byte) Token {
+    // the partition key is the Token
+    return orderedToken(partitionKey)
 }
 
-func (p orderedPartitioner) ParseString(str string) token {
-	return orderedToken(str)
+func (p orderedPartitioner) ParseString(str string) Token {
+    return orderedToken(str)
 }
 
 func (o orderedToken) String() string {
-	return string(o)
+    return string(o)
 }
 
-func (o orderedToken) Less(token token) bool {
-	return o < token.(orderedToken)
+func (o orderedToken) Less(token Token) bool {
+    return o < token.(orderedToken)
 }
 
 // random partitioner and token
 type randomPartitioner struct{}
 type randomToken big.Int
 
 func (r randomPartitioner) Name() string {
-	return "RandomPartitioner"
+    return "RandomPartitioner"
 }
 
 // 2 ** 128
 var maxHashInt, _ = new(big.Int).SetString("340282366920938463463374607431768211456", 10)
 
-func (p randomPartitioner) Hash(partitionKey []byte) token {
-	sum := md5.Sum(partitionKey)
-	val := new(big.Int)
-	val.SetBytes(sum[:])
-	if sum[0] > 127 {
-		val.Sub(val, maxHashInt)
-		val.Abs(val)
-	}
+func (p randomPartitioner) Hash(partitionKey []byte) Token {
+    sum := md5.Sum(partitionKey)
+    val := new(big.Int)
+    val.SetBytes(sum[:])
+    if sum[0] > 127 {
+        val.Sub(val, maxHashInt)
+        val.Abs(val)
+    }
 
-	return (*randomToken)(val)
+    return (*randomToken)(val)
 }
 
-func (p randomPartitioner) ParseString(str string) token {
-	val := new(big.Int)
-	val.SetString(str, 10)
-	return (*randomToken)(val)
+func (p randomPartitioner) ParseString(str string) Token {
+    val := new(big.Int)
+    val.SetString(str, 10)
+    return (*randomToken)(val)
 }
 
 func (r *randomToken) String() string {
-	return (*big.Int)(r).String()
+    return (*big.Int)(r).String()
 }
 
-func (r *randomToken) Less(token token) bool {
-	return -1 == (*big.Int)(r).Cmp((*big.Int)(token.(*randomToken)))
+func (r *randomToken) Less(token Token) bool {
+    return -1 == (*big.Int)(r).Cmp((*big.Int)(token.(*randomToken)))
 }
 
 type hostToken struct {
-	token token
-	host  *HostInfo
+    token Token
+    host  *HostInfo
 }
 
 func (ht hostToken) String() string {
-	return fmt.Sprintf("{token=%v host=%v}", ht.token, ht.host.HostID())
+    return fmt.Sprintf("{token=%v host=%v}", ht.token, ht.host.HostID())
 }
 
-// a data structure for organizing the relationship between tokens and hosts
-type tokenRing struct {
-	partitioner partitioner
-
-	// tokens map token range to primary replica.
-	// The elements in tokens are sorted by token ascending.
-	// The range for a given item in tokens starts after preceding range and ends with the token specified in
-	// token. The end token is part of the range.
-	// The lowest (i.e. index 0) range wraps around the ring (its preceding range is the one with largest index).
-	tokens []hostToken
+// TokenRing is a data structure for organizing the relationship between tokens and hosts
+type TokenRing interface {
```

Comment:
**suggestion**: Don't export interface. Interfaces should usually be defined at the user side of an API. Adding a method to interface is a breaking change unless there is a private method that guarantees that users can't provide their own implementation. Also, we won't be providing multiple TokenRing implementations. So it seems better to export a pointer to struct directly.

Same with ClusterMetadata.